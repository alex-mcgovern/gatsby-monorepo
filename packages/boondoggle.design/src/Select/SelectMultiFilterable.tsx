import React from "react";
import type { DropdownItem } from "../..";
import { Box } from "../Box";
import type { InputProps } from "../Input";
import { Label } from "../Label";
import type { GetSprinklesArgs } from "../__css__/getSprinkles.css";
import { DropdownMenu } from "./components/DropdownMenu";
import { SelectInput } from "./components/SelectInput";
import { DownshiftProviderMulti } from "./context/DownshiftProviderMulti";

export interface SelectMultiFilterableProps
  extends Omit<InputProps, "value" | "iconTrailing">,
    GetSprinklesArgs {
  items: Array<DropdownItem>;
  initialValue?: Array<DropdownItem>;
  onValueChange?(...args: Array<unknown>): unknown;
  onSelectItem?(...args: Array<unknown>): unknown;
  selectedItems?: Array<DropdownItem>;
  label?: string;
}

export function SelectMultiFilterable({
  items,
  onValueChange,
  onSelectItem,
  initialValue,
  id,
  disabled,
  iconLeading,
  label,
  placeholder,
  size,
  name,
  ...rest
}: SelectMultiFilterableProps) {
  return (
    <DownshiftProviderMulti
      items={items}
      onValueChange={onValueChange}
      onSelectItem={onSelectItem}
      initialValue={initialValue}
    >
      <Box position="relative" width="100%" {...rest}>
        {label && <Label htmlFor={id} label={label} />}
        <SelectInput
          disabled={disabled}
          iconLeading={iconLeading}
          id={id}
          name={name}
          placeholder={placeholder}
          size={size}
        />
        <DropdownMenu size={size} isSelectMultiFilterable />
      </Box>
    </DownshiftProviderMulti>
  );
}
