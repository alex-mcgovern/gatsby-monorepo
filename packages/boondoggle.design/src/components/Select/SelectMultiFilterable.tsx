import React from "react";
import type { GetSprinklesArgs } from "../../styles/getSprinkles.css";
import { Box } from "../Box";
import type { InputProps } from "../Input";
import { Label } from "../Label";
import { DropdownMenu } from "./components/DropdownMenu";
import { SelectInput } from "./components/SelectInput";
import { DownshiftProviderMulti } from "./context/DownshiftProviderMulti";
import type { DropdownItem } from "./types";

export interface SelectMultiFilterableProps
  extends Omit<InputProps, "value" | "iconTrailing">,
    GetSprinklesArgs {
  items: Array<DropdownItem>;
  initialValue?: Array<DropdownItem>;
  onValueChange?(...args: Array<unknown>): unknown;
  onSelectItem?(...args: Array<unknown>): unknown;
  selectedItems?: Array<DropdownItem>;
  label?: string;
}

export function SelectMultiFilterable({
  items,
  onValueChange,
  onSelectItem,
  initialValue,
  id,
  disabled,
  iconLeading,
  label,
  placeholder,
  size,
  name,
  ...rest
}: SelectMultiFilterableProps) {
  return (
    <DownshiftProviderMulti
      items={items}
      onValueChange={onValueChange}
      onSelectItem={onSelectItem}
      initialValue={initialValue}
    >
      <Box position="relative" width="100%" {...rest}>
        {label && <Label htmlFor={id} label={label} />}
        <SelectInput
          disabled={disabled}
          iconLeading={iconLeading}
          id={id}
          name={name}
          placeholder={placeholder}
          size={size}
        />
        <DropdownMenu size={size} isSelectMultiFilterable />
      </Box>
    </DownshiftProviderMulti>
  );
}
