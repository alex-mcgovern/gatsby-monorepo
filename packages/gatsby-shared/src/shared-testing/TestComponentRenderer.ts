import type { JSXElementConstructor, ReactElement } from "react";
import { cloneElement } from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

export const renderTestComponent = <TComponentProps>(
  Component: ReactElement<JSXElementConstructor<TComponentProps>>
) => {
  const user = userEvent.setup();

  const { props } = Component;

  const propertiesWithTestId = {
    ...props,
    "data-testid": "tested-component",
  };

  const result = render(cloneElement(Component, propertiesWithTestId));

  const component = screen.queryByTestId("tested-component");

  return {
    ...result,
    user,
    component: component && component,
  };
};
