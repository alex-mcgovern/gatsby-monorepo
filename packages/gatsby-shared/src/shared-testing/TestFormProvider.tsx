import type { JSXElementConstructor, ReactElement } from "react";
import React, { cloneElement } from "react";
import { FormProvider, useForm } from "react-hook-form";

export function FormTestProvider({
  children,
  ...rest
}: {
  children: ReactElement<unknown, string | JSXElementConstructor<unknown>>;
}) {
  const reactHookFormMethods = useForm();

  return (
    <FormProvider {...reactHookFormMethods}>
      {cloneElement(children, {
        ...rest,
      })}
    </FormProvider>
  );
}
