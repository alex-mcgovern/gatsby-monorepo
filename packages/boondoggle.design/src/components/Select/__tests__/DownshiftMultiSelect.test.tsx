/** @jest-environment jsdom */
import React from "react";
import { renderTestComponent } from "@alexmcgovern/gatsby-shared/src/shared-testing/TestComponentRenderer";
import { cleanup } from "@testing-library/react";
import type { SelectMultiFilterableProps } from "../SelectMultiFilterable";
import { SelectMultiFilterable } from "../SelectMultiFilterable";

afterEach(cleanup);

test("Renders without error", async () => {
  const { getByTestId } = renderTestComponent<SelectMultiFilterableProps>(
    <SelectMultiFilterable
      id="test-dropdown"
      items={[
        { value: "foo", label: "foo" },
        { value: "bar", label: "bar" },
        { value: "bob", label: "bob" },
        { value: "alice", label: "alice" },
      ]}
      name="test-dropdown"
      size={undefined}
    />
  );

  expect(getByTestId("tested-component")).not.toBeNull();
  expect(getByTestId("tested-component")).toMatchSnapshot();
});
