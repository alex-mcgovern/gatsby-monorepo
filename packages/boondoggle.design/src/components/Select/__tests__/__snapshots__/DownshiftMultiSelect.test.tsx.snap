// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders without error 1`] = `
<div
  class="getSprinkles_position_relative__1jganc99z getSprinkles_width_100%_mobile__1jganc9e4"
  data-testid="tested-component"
>
  <div>
    <fieldset
      class="Input_getInputWrapperStyles__17v4xld1 Input__17v4xld0 getSprinkles_border_neutral_border_interactive__1jganc92o getSprinkles_borderRadius_md__1jganc92i getSprinkles_background_neutral_background_base__1jganc9x getSprinkles_color_neutral_text_lowContrast__1jganc922 getSprinkles_display_flex__1jganc9c getSprinkles_alignItems_center__1jganc92 getSprinkles_width_100%_mobile__1jganc9e4 getSprinkles_overflow_hidden__1jganc989 Input_getInputWrapperStyles_size_md__17v4xld3 variant_interactive_element_size_variantInteractiveElementSize_md__t3r2od7 variant_interactive_element_size__t3r2od1 getSprinkles_fontSize_body_lg_mobile__1jganc9ay getSprinkles_fontSize_body_md_desktop__1jganc9b3 getSprinkles_lineHeight_body_lg_mobile__1jganc9bs getSprinkles_lineHeight_body_md_desktop__1jganc9bx getSprinkles_height_spacing5_mobile__1jganc9hm getSprinkles_height_spacing4_desktop__1jganc9hl focus_ring_styles_focusedStateStyle__1vv8g8r0 "
      role="combobox"
    >
      <input
        aria-activedescendant=""
        aria-autocomplete="list"
        aria-controls="downshift-0-menu"
        aria-expanded="false"
        aria-labelledby="downshift-0-label"
        autocomplete="off"
        class="Input_inputElement__17v4xld9 Input__17v4xld8 getSprinkles_paddingLeft_spacing1__1jganc990 getSprinkles_paddingRight_spacing1__1jganc99f getSprinkles_width_100%_mobile__1jganc9e4 getSprinkles_height_100%_mobile__1jganc9gj"
        id="test-dropdown"
        name="test-dropdown"
        placeholder="All"
        value=""
      />
      <div>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-angle-down icon_icon__111gfxw0 Input_trailingIcon__17v4xldd Input__17v4xldc getSprinkles_marginRight_spacing2__1jganc95o"
          data-icon="angle-down"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M201.4 374.6c12.5 12.5 32.8 12.5 45.3 0l160-160c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 306.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l160 160z"
            fill="currentColor"
          />
        </svg>
      </div>
    </fieldset>
  </div>
  <div
    aria-labelledby="downshift-0-label"
    class="DropdownMenu_dropdownMenu__1ubprtm1 focus_ring_styles_focusedStateStyle__1vv8g8r0 DropdownMenu__1ubprtm0 getSprinkles_marginTop_spacing1__1jganc962 getSprinkles_width_max-content_mobile__1jganc9ej getSprinkles_maxWidth_gridSpan3__1jganc96v getSprinkles_display_block__1jganc9b getSprinkles_margin_none__1jganc93x getSprinkles_borderRadius_sm__1jganc92j getSprinkles_overflow_hidden__1jganc989 getSprinkles_background_neutral_background_raised__1jganc9y getSprinkles_boxShadow_sm__1jganc9a getSprinkles_border_neutral_border_interactive__1jganc92o getSprinkles_overflowY_auto__1jganc98a DropdownMenu_dropdownMenuClosed__1ubprtm4"
    id="downshift-0-menu"
    role="listbox"
  >
    <div
      class="DropdownMenu_dropdownListInner__1ubprtm3 DropdownMenu__1ubprtm2"
    >
      <span
        aria-selected="false"
        class="DropdownListItem_getDropdownListItemStyles__mbnr2x1 DropdownListItem__mbnr2x0 getSprinkles_paddingLeft_spacing1__1jganc990 getSprinkles_paddingRight_spacing1__1jganc99f getSprinkles_display_flex__1jganc9c getSprinkles_alignItems_center__1jganc92 getSprinkles_textDecoration_none__1jganc9a6 getSprinkles_width_100%_mobile__1jganc9e4 getSprinkles_textAlign_left__1jganc9a3 getSprinkles_color_neutral_text_highContrast__1jganc923 getSprinkles_whiteSpace_nowrap__1jganc9aa getSprinkles_margin_none__1jganc93x getSprinkles_gap_spacing0__1jganc93d focus_ring_styles_focusedStateStyle__1vv8g8r0 disabled_state_styles_disabledStateStyles__5jo1bk0 DropdownListItem_getDropdownListItemStyles_size_md__mbnr2x3 variant_interactive_element_size_variantInteractiveElementSize_md__t3r2od7 variant_interactive_element_size__t3r2od1 getSprinkles_fontSize_body_lg_mobile__1jganc9ay getSprinkles_fontSize_body_md_desktop__1jganc9b3 getSprinkles_lineHeight_body_lg_mobile__1jganc9bs getSprinkles_lineHeight_body_md_desktop__1jganc9bx getSprinkles_height_spacing5_mobile__1jganc9hm getSprinkles_height_spacing4_desktop__1jganc9hl"
        id="downshift-0-item-0"
        role="option"
      >
        <input
          readonly=""
          tabindex="-1"
          type="checkbox"
        />
        foo
      </span>
      <span
        aria-selected="false"
        class="DropdownListItem_getDropdownListItemStyles__mbnr2x1 DropdownListItem__mbnr2x0 getSprinkles_paddingLeft_spacing1__1jganc990 getSprinkles_paddingRight_spacing1__1jganc99f getSprinkles_display_flex__1jganc9c getSprinkles_alignItems_center__1jganc92 getSprinkles_textDecoration_none__1jganc9a6 getSprinkles_width_100%_mobile__1jganc9e4 getSprinkles_textAlign_left__1jganc9a3 getSprinkles_color_neutral_text_highContrast__1jganc923 getSprinkles_whiteSpace_nowrap__1jganc9aa getSprinkles_margin_none__1jganc93x getSprinkles_gap_spacing0__1jganc93d focus_ring_styles_focusedStateStyle__1vv8g8r0 disabled_state_styles_disabledStateStyles__5jo1bk0 DropdownListItem_getDropdownListItemStyles_size_md__mbnr2x3 variant_interactive_element_size_variantInteractiveElementSize_md__t3r2od7 variant_interactive_element_size__t3r2od1 getSprinkles_fontSize_body_lg_mobile__1jganc9ay getSprinkles_fontSize_body_md_desktop__1jganc9b3 getSprinkles_lineHeight_body_lg_mobile__1jganc9bs getSprinkles_lineHeight_body_md_desktop__1jganc9bx getSprinkles_height_spacing5_mobile__1jganc9hm getSprinkles_height_spacing4_desktop__1jganc9hl"
        id="downshift-0-item-1"
        role="option"
      >
        <input
          readonly=""
          tabindex="-1"
          type="checkbox"
        />
        bar
      </span>
      <span
        aria-selected="false"
        class="DropdownListItem_getDropdownListItemStyles__mbnr2x1 DropdownListItem__mbnr2x0 getSprinkles_paddingLeft_spacing1__1jganc990 getSprinkles_paddingRight_spacing1__1jganc99f getSprinkles_display_flex__1jganc9c getSprinkles_alignItems_center__1jganc92 getSprinkles_textDecoration_none__1jganc9a6 getSprinkles_width_100%_mobile__1jganc9e4 getSprinkles_textAlign_left__1jganc9a3 getSprinkles_color_neutral_text_highContrast__1jganc923 getSprinkles_whiteSpace_nowrap__1jganc9aa getSprinkles_margin_none__1jganc93x getSprinkles_gap_spacing0__1jganc93d focus_ring_styles_focusedStateStyle__1vv8g8r0 disabled_state_styles_disabledStateStyles__5jo1bk0 DropdownListItem_getDropdownListItemStyles_size_md__mbnr2x3 variant_interactive_element_size_variantInteractiveElementSize_md__t3r2od7 variant_interactive_element_size__t3r2od1 getSprinkles_fontSize_body_lg_mobile__1jganc9ay getSprinkles_fontSize_body_md_desktop__1jganc9b3 getSprinkles_lineHeight_body_lg_mobile__1jganc9bs getSprinkles_lineHeight_body_md_desktop__1jganc9bx getSprinkles_height_spacing5_mobile__1jganc9hm getSprinkles_height_spacing4_desktop__1jganc9hl"
        id="downshift-0-item-2"
        role="option"
      >
        <input
          readonly=""
          tabindex="-1"
          type="checkbox"
        />
        bob
      </span>
      <span
        aria-selected="false"
        class="DropdownListItem_getDropdownListItemStyles__mbnr2x1 DropdownListItem__mbnr2x0 getSprinkles_paddingLeft_spacing1__1jganc990 getSprinkles_paddingRight_spacing1__1jganc99f getSprinkles_display_flex__1jganc9c getSprinkles_alignItems_center__1jganc92 getSprinkles_textDecoration_none__1jganc9a6 getSprinkles_width_100%_mobile__1jganc9e4 getSprinkles_textAlign_left__1jganc9a3 getSprinkles_color_neutral_text_highContrast__1jganc923 getSprinkles_whiteSpace_nowrap__1jganc9aa getSprinkles_margin_none__1jganc93x getSprinkles_gap_spacing0__1jganc93d focus_ring_styles_focusedStateStyle__1vv8g8r0 disabled_state_styles_disabledStateStyles__5jo1bk0 DropdownListItem_getDropdownListItemStyles_size_md__mbnr2x3 variant_interactive_element_size_variantInteractiveElementSize_md__t3r2od7 variant_interactive_element_size__t3r2od1 getSprinkles_fontSize_body_lg_mobile__1jganc9ay getSprinkles_fontSize_body_md_desktop__1jganc9b3 getSprinkles_lineHeight_body_lg_mobile__1jganc9bs getSprinkles_lineHeight_body_md_desktop__1jganc9bx getSprinkles_height_spacing5_mobile__1jganc9hm getSprinkles_height_spacing4_desktop__1jganc9hl"
        id="downshift-0-item-3"
        role="option"
      >
        <input
          readonly=""
          tabindex="-1"
          type="checkbox"
        />
        alice
      </span>
    </div>
  </div>
</div>
`;
